text/0000-transaction-refactoring.md

```diff
@@ -0,0 +1,327 @@
+
+`Transaction` contains the following fields
+- `originator_id` is an account ID of the transaction originator.
+- `public_key` is to identify the access key used to signs the transaction.
+- `nonce` is used to deduplicate and order transactions.
+- `receiver_id` is where the transaction has to be routed.
+- `action` is the list of actions to perform.
+
+An `Action` can be of the following:
+- `CreateAccount` creates a new account. It has to be the first action, and the receiver has to be a new account. The action will fail if the account already exists.
+`CreateAccount` also grants permission for all subsequent batched action for the newly created account. For example, deploying code on the new account.
+- `DeployContract` deploys given binary wasm code on the account. Either the `receiver_id` equals to the `originator_id`, or the batch of actions has started with `CreateAccount`, which granted that permission.
+- `FunctionCall` executes a function call on the last deployed contract. E.g. if the previous action was `DeployContract`, then the code to execute will be the new deployed contract. `FunctionCall` has `method_name` and `args` to identify method with arguments to call. It also has `fee` and the `amount`. `fee` is the transaction fee that is prepaid for this call to be spent on gas. `amount` is the attached amount of NEAR tokens that the contract can spend, e.g. 10 tokens to pay for a crypto-corgi.
+- `Transfer` transfers the given amount of tokens from the originator to the receiver.
+- `Stake` stakes the new total amount which is given in the `amount` field with the given `public_key`. The difference in stake is taken from the account's balance at the moment when this action is executed, so it's not prepaid. There is no particular reason to stake on behalf of a newly created account, so we may disallow it.
+- `DeleteKey` deletes an old `AccessKey` identified by the given `public_key` from the account. Fails if the access key with the given public key doesn't exist. All next batched actions will continue to execute, even if the public key that authorized that transaction was removed.
+- `AddKey` adds a new given `AccessKey` identified by a new given `public_key` to the account. Fails if an access key with the given public key already exists. We removed `SwapKeyTransaction`, because it can be replaced with 2 batched actions - delete an old key and add a new key.
+
+The new `Receipt` contains the shared information and either one of the receipt actions or a list of actions:
+- `sender_id` the account ID of the immediate previous sender of this receipt. It can be different from the `originator_id` in some cases.
+- `receiver_id` the account ID of the current account, on which we need to perform action(s).
+- `receipt_id` is a hash of this receipt (previously was called `nonce`). It's generated from either the signed transaction or the parent receipt.
```

Comment:
It's not necessarily the hash of this receipt. It's better to call it something like "unique identifier".

Reply:
right.